src/background/legacy-external-message-handler.ts

```diff
@@ -0,0 +1,133 @@
+import { formatAuthResponse } from '@shared/actions/finalize-auth-response';
+import { formatMessageSigningResponse } from '@shared/actions/finalize-message-signature';
+import { formatTxSignatureResponse } from '@shared/actions/finalize-tx-signature';
+import {
+  ExternalMethods,
+  LegacyMessageFromContentScript,
+  LegacyMessageToContentScript,
+} from '@shared/message-types';
+import { RouteUrls } from '@shared/route-urls';
+import { popupCenter } from './popup-center';
+
+const IS_TEST_ENV = process.env.TEST_ENV === 'true';
+
+//
+// Playwright does not currently support Chrome extension popup testing:
+// https://github.com/microsoft/playwright/issues/5593
+async function openRequestInFullPage(path: string, urlParams: URLSearchParams) {
+  return chrome.tabs.create({
+    url: chrome.runtime.getURL(`index.html#${path}?${urlParams.toString()}`),
+  });
+}
+
+export function inferLegacyMessage(message: any): message is LegacyMessageFromContentScript {
+  // Now that we use a RPC communication style, we can infer
+  // legacy message types by presence of an id
+  const hasIdProp = 'id' in message;
+  return !hasIdProp;
+}
+
+function getTabIdFromPort(port: chrome.runtime.Port) {
+  return port.sender?.tab?.id;
+}
+
+function getOriginFromPort(port: chrome.runtime.Port) {
+  return port.sender?.origin;
+}
+
+function makeSearchParamsWithDefaults(
+  port: chrome.runtime.Port,
+  otherParams: [string, string][] = []
+) {
+  const urlParams = new URLSearchParams();
+  // All actions must have a corresponding `origin` and `tabId`
+  const origin = getOriginFromPort(port);
+  const tabId = getTabIdFromPort(port);
+  urlParams.set('origin', origin ?? '');
+  urlParams.set('tabId', tabId?.toString() ?? '');
+  otherParams.forEach(([key, value]) => urlParams.set(key, value));
+  return { urlParams, origin, tabId };
+}
+
+interface ListenForWindowCloseArgs {
+  // ID that comes from newly created window
+  id?: number;
```

Comment:
Could these be named in a way that you don't need to provide comments?

Reply:
yes, but then we'll need to rename from id --> descriptive in each case (as default name is `id`)